org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/preferences/internal/DefaultModelPersistence.java

```diff
@@ -51,34 +53,37 @@
  */
 public final class DefaultModelPersistence implements ModelPersistence, EventListener {
 
-    // Cache holding (project name - persistent model property) pairs.
-    private final Map<IProject, PersistentModel> cache;
+    private final LoadingCache<IProject, PersistentModel> modelCache;
 
     private DefaultModelPersistence() {
-        this.cache = Maps.newConcurrentMap();
+        this.modelCache = CacheBuilder.newBuilder().build(new CacheLoader<IProject, PersistentModel>() {
+
+            @Override
+            public PersistentModel load(IProject project) throws Exception {
+                return doLoadModel(project);
+            }
+        });
     }
 
     @Override
     public PersistentModel loadModel(IProject project) {
-        PersistentModel model = this.cache.get(project);
-        if (model == null) {
-            model = readModel(project);
-            if (model != null) {
-                this.cache.put(project, model);
-            }
+        PersistentModel model = this.modelCache.getUnchecked(project);
+        if (model instanceof AbsentPersistentModel) {
+            return null;
```

Comment:
I'm totally lost now. Two iterations ago you this comment: 

> The current solution would silently hide programming errors (forgetting to check isEmptyModel).

and

> By default, null should be returned.

Frankly, I think both of them are equally right, and it's only a syntactic sugar whether to check for `null` or for `isPresent` models. So, `isPresent()` is your preference?